content/subtleties.md

```diff
@@ -14,37 +14,39 @@ Also see Pierre Weis's [Frequently asked Questions about Caml](http://caml.inria
+  
+  The value restriction is a pessimistic assumption that a compiler must make in the face of the possible mutability, and is a price that must be payed even if mutability is not used. Weak type variables alleviate the problem a little, by allowing us to omit type annotations for values that don't have concrete or polymorphic type. However, since types are analyzed on the compilation unit level (the file level), weak type variables can't escape the scope of a compilation unit (the scope of a file). So it is not possible to have a weak type variable in the module interface (e.g., in the `mli` file). Compiler will reject all toplevel definitions that have weak types, unless they are either hidden from the module interface (i.e., omitted in the interface - e.g., when the mli file is empty) or given a concrete type. 
+
+  If for some reason compiler refuses to generalize your expression and weak type variables occur in your type, then you can try the following options to fix the situation:
+  - Convert your expression to a syntactic constant, by making it a function, aka eta-expansion. E.g., `let x = ref None` could be made polymorphic by making `x` a function, e.g., `let x () = ref None` 
+  - If the problematic type is abstract, then a compiler can't perform the variance analysis on it. You can expose, that some type variables of your abstract type are covariant, by making the variance annotations, e.g., `type +'a t` in the signature will tell the compiler that values of that type could be safely given the polymorphic type. (Of course, the definition should correspond to the specification, that is checked by the compiler, for example, `type +'a t = 'a list` would be accepted by the type checker, where `type +'a t = 'a ref` will be rejected). 
+  - If nothing else helps, then you need to manually annotate your value with a ground type or hide from the interface. 
```

Comment:
Give an example of a "ground type". Also, what is the referent of "hide from the interface"? Perhaps "hide _it_ from the interface" so it refers to "your value"?

Reply:
btw if you guys feel it's reached a fixed point before I get a chance to look at it seriously later today, I will certainly have no objection to publishing whatever you've done.  I certainly don't feel ownership of the text, and am grateful that you three are working on it.  I didn't have a deep understanding of the issues; I wrote this to get something started.  I can always suggest refinements later if I think there's something important.

Reply:
I do like the idea of having a short, informal text, and then more detailed text that's separate--perhaps on a different page that's linked from the short text.